### ISSUE TITLE
[(ngModel)] binding to object gets corrupted

### ISSUE THREAD

**Issue by corner22** [ng2-ui/ng2-auto-complete#43]
I am using ng-auto-complete v0.11.4, with angular 2.0.0, as follows:
```
<input ng2-auto-complete name="row{{row.DisplayOrder}}Task2" 
[source]="getTaskListFunction(row)"
[(ngModel)]="row.TaskListItem"
min-chars="2"
[list-formatter]="taskAutoCompleteFormatter"
value-property-name="Id"
display-property-name="Value"
/>
```
I am defining this element inside a table of repeating tr rows using *ngFor.  The table is contained within a form.

getTaskListFunction() returns a function, and this function, in turn, returns an `Observable<TimesheetTaskListItem[]>`.

row.TaskListItem is an instance of TimesheetTaskListItem object, defined as

`class TimesheetTaskListItem {
    Id: number;
    Value: string;
}`

Everything works great, up to the point where I submit my form (calling WebAPI).  Until I submit the form, the two-way binding to row.TaskListItem works as expected in that when first rendering the input element it shows the 'Value' property of the bound object, and when I select a different object from the auto-complete list it will correctly replace the instance referenced by row.TaskListItem with the object selected in the list and display the 'Value' property of this new object.  

The binding gets screwed once I submit the form (to a WebAPI operation).  After making this submission, the binding displays the contents of the input box as "[object Object]", which I interpret to mean that it is no longer respecting the display-property-name attribute.  If I reload the data that my form is bound to then the binding remains corrupted, even if I load data for a different record.

Note that when I submit the form, I do not modify the row.TaskListItem object in any way and I do not modify any attributes of the input element.

I have tried my best to isolate where in my code or auto-complete's code this corruption first occurs but I am unable to diagnose any further than saving it is occurring at some point during/after the form submission.  As for the resolution, I can only assume that the fix lies:

1. I'm doing something wrong in my definition of input element.
2. There is something wrong with the two-way binding in ng2-auto-complete.
3. This is a more general bug with angular2.  

Have you got any ideas about the cause of this problem, and if not a fix, then any ideas on a workaround.

Thanks in advance

**Comment by allenhwkim** [comment]
Thanks for reporting issue although it's an unknown issue.

Because I do not see your application, I cannot pinpoint the cause of your issue. However, based on  what I read, I  try to list the following; Please ignore any if it's not your case

1. `[source]` should return an Observable, not Observable.row

2. `[source]` as an Observable should take an argument, which is user input text as string. As you see in the example, `[source]="appsvc.findHeros"` and it looks like `findHeros(arg) { ... }`

3. For your SPA, take a look at `rows` -> `row` -> `rowTaskListItem`, and investigate how `rows` changes. 

4. Use `(valueChanged)`  or `(ngModelChange)` to look at what value are actually assigned to your ngModel. e.g. `(ngModelChange)="myFunc($event)`

**Comment by corner22** [comment]
Hi Allen.

The bit about the source function was tricky to explain, so sorry if I wasn't clear, but it does ultimately return a function with a return type of Observable and the function takes a single string parameter of the text to search for.  This part of the auto-complete mechanism works  To show you the function:

```
    taskListService : any = {};
 
    getTaskListFunction(row: TimesheetRowViewModel): any {
        let deptId = row.DeptId;
        let taskId = row.TaskListItem.Id;
        let timesheetService = this._timesheetService;
        const functionName = "getTaskList_" + (!!deptId ? deptId : 0).toString() + "_" + (!!taskId ? taskId : 0).toString();
        if (!this.taskListService[functionName]) {
            this.taskListService[functionName] = function (textToMatch: string): Observable<any[]> {
                return timesheetService.getFilteredTasks(deptId, taskId, textToMatch);
            };
        }
        return this.taskListService[functionName];
    }
```

I have experimented with (valueChanged) or (ngModelChange) as well as debugging in Chrome and I am 100% positive that the row.TaskListItem is being handled correctly by my peripheral code.

What I have noticed since my last post though is an anomaly with the binding that may or may not be relevant....  If I put a breakpoint inside my getTaskListFunction function I can see that when I start typing into the input box the value of row.TaskListItem is set to the string value that I have typed into the input control and is no longer an instance of the TimesheetTaskListItem class that I expect it to be.  Before I start typing, and after I have selected from the autocomplete list the value of row.TaskListItem is, correctly, set to an instance of TimesheetTaskListItem.

So it seems that when typing into the input control the autocomplete code is somehow setting the bound property to the string text that has been typed, and is not respecting the display-property-name attribute.  In other words, I would expect that any value typed into the input control would be written to the row.TaskListItem.Value property rather than being written to the row.TaskListItem property.  

I hope I have made myself clear.

Thanks

**Comment by corner22** [comment]
I think we can now consider this issue closed. Unfortunately, I can't give a concrete reason to why it works for me now.  I made a couple of changes and one, or both, have resolved the issue.

1. I discovered that I had installed ng2-ui 0.11.4 which in turn was loading v0.6.6 of ng2-auto-complete.  I could see that there was a later version of (0.7.4) of ng2-auto-complete so I upgraded this component.  Because of other work-in-progress my usage of ng2-auto-complete was broken at the time so I wasn't able to prove that upgrading to 0.7.4 resolved the issue.

2. I noticed that there was a new release of angular2 (2.1.1) so I upgraded this from 2.0.1, and all dependencies, to latest version.

once I had fixed my broken work-in-progress code I discovered that the auto-complete binding was now working as expected.  It was no longer overwriting row.taskListitem with a string and the binding-state was no longer being corrupted after a form submission,

Only wish I could tell you which of the two upgrades resolved my issue.
